### PR TITLE
[SEC-180] Add ability to overlay classes with new definitions of existing methods

### PR DESCRIPTION
**Pull request by NathanJPhillips**
Allows multiple definitions of the same class to appear on the classpath, so long as all but the first definition are marked with the attribute `@java::com.diffblue.security.OverlayClassImplementation`. Overlay class definitions can contain methods with the same signature as methods in the original class, so long as these are marked with the attribute `@java::com.diffblue.security.OverlayMethodImplementation`; such overlay methods are replaced in the original file with the version found in the last overlay on the classpath. Later definitions can also contain new supporting methods and fields that are merged in.
This will allow insertion of Java methods into library classes to handle, for example, modelling dependency injection.

### CODE REVIEW

**Comment by tautschnig** [comment]
Would it be possible to have a more complete description of what this is supposed to do? I am struggling to parse "Add ability to overlay classes with new definitions of existing methods", because "classes" and "methods" doesn't really seem to fit together?

**Comment by NathanJPhillips** [comment]
@tautschnig absolutely. Thank you for asking.

**Comment by tautschnig** [comment]
Thanks a lot for amending the description, that's very helpful! Would you mind copying the same to the commit message? You might very well have done this already, and just not pushed -- there is of course not rush, I'd just hope to have it in the very final version.

**Comment by NathanJPhillips** [comment]
I had a vague memory that commit messages were what and PR descriptions were why so I hadn't done this. It is however in a comment in class_loader.cpp. Will update when I do my next push.

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 1]
```diff
@@ -48,42 +48,90 @@ class java_bytecode_convert_classt:public messaget
   {
   }
 
-  void operator()(const java_bytecode_parse_treet &parse_tree)
+  /// Converts a class parse tree into a class symbol and adds it to the
+  /// symbol table.
+  /// \param parse_trees: The parse trees found for the class to be converted.
+  /// \remarks
+  ///   Allows multiple definitions of the same class to appear on the
+  ///   classpath, so long as all but the first definition are marked with the
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: This implies if the first is annotated with `OverlayClassImplementation` this won't work, is that correct/desirable (it seems like if exactly one are marked with the `OverlayClassImplementation` tag as the ordering of loading classes doesn't seem like an ideal thing to depend on.

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 2]
```diff
@@ -48,42 +48,90 @@ class java_bytecode_convert_classt:public messaget
...
-    bool loading_success=parse_tree.loading_successful;
-    if(loading_success)
-      convert(parse_tree.parsed_class);
+    const java_bytecode_parse_treet &parse_tree = parse_trees.front();
+    const irep_idt &class_name = parse_tree.parsed_class.name;
+    bool loading_success = parse_tree.loading_successful;
+    // Check that the first class implementation isn't an overlay
+    if(is_overlay_class(parse_tree.parsed_class))
+    {
+      warning()
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: OK so yeah the comment is correct - perhaps a task for a future day to allow the classes to come in any order

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 3]
```diff
@@ -48,42 +48,90 @@ class java_bytecode_convert_classt:public messaget
...
+        << "Skipping class " << class_name << " marked with "
+        << "OverlayClassImplementation but found before original definition"
+        << eom;
+    }
+    else if(loading_success)
+    {
+      overlay_classest overlay_classes;
+      for(auto parse_tree_it = std::next(parse_trees.begin());
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]:
```c++
std::filter(
  std::next(parse_trees.begin()), 
  parse_trees.end(), 
  std::front_inserter(overlay_classes), 
  [](parse_tree) -> { return parse_tree.loading_succesful && is_overlay_class(parse_tree); });
```
?

(you can probably still get the warning in inside the lambda, or could separately find mutliple non-overlay classes.

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 4]
```diff
@@ -292,20 +337,79 @@ void java_bytecode_convert_classt::convert(const classt &c)
...
-  // now do fields
+  // Now do fields
+  for(auto overlay_class : overlay_classes)
+  {
+    for(const auto &field : overlay_class.get().fields)
+    {
+      debug()
+        << "Adding symbol from overlay class:  field '" << field.name << "'"
+        << eom;
+      convert(*class_symbol, field);
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: So all the fields from the overlay classes are added to the original? This might benefit from a comment

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 5]
```diff
@@ -292,20 +337,79 @@ void java_bytecode_convert_classt::convert(const classt &c)
...
+    }
+  }
   for(const auto &field : c.fields)
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: I would add the "real" fields first, then if an overlay field is going to hide an existing field can drop a warning, unless there is a good reason not to.

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 6]
```diff
@@ -292,20 +337,79 @@ void java_bytecode_convert_classt::convert(const classt &c)
...
+  std::set<irep_idt> overlay_methods;
+  for(auto overlay_class : overlay_classes)
+  {
+    for(const methodt &method : overlay_class.get().methods)
+    {
+      const irep_idt method_identifier =
+        id2string(qualified_classname) + "." + id2string(method.name)
+          + ":" + method.descriptor;
+      if(method_bytecode.contains_method(method_identifier))
+      {
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: Does this case not deserve a warning?

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 7]
```diff
@@ -292,20 +337,79 @@ void java_bytecode_convert_classt::convert(const classt &c)
...
+        if(overlay_methods.count(method_identifier) == 0)
+        {
+          warning()
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: I don't understand this warning. The conditions above as I understand it say:

 - we're looking at a method from an overlay class 
 - the method also exists in the original class
 - the method has not already been seen on any overlay for this specific class

But your warning seems to suggest:
 - is_overlay_method returned false 
 - and it _has_ been seen already i.e. `overlay_methods.count(method_identifier) > 0` 

Not sure which is correct but I think something isn't matching up here.

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 8]
```diff
@@ -292,20 +337,79 @@ void java_bytecode_convert_classt::convert(const classt &c)
...
+        overlay_methods.insert(method_identifier);
+    }
+  }
+  for(const methodt &method : c.methods)
   {
     const irep_idt method_identifier=
       id2string(qualified_classname)+
       "."+id2string(method.name)+
       ":"+method.descriptor;
+    if(method_bytecode.contains_method(method_identifier))
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: This is false when we've added a overlay method from a class? Perhaps a comment here

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 9]
```diff
@@ -292,20 +337,79 @@ void java_bytecode_convert_classt::convert(const classt &c)
...
+    {
+      if(overlay_methods.erase(method_identifier) == 0)
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: Feels like this might even be an invariant?

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 10]
```diff
@@ -316,6 +420,21 @@ void java_bytecode_convert_classt::convert(const classt &c)
       symbol_table,
       get_message_handler());
     method_bytecode.add(qualified_classname, method_identifier, method);
+    if(is_overlay_method(method))
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp:540]: Again this feels like an invariant?

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 11]
```diff
@@ -592,8 +711,41 @@ void java_bytecode_convert_classt::add_array_types(symbol_tablet &symbol_table)
...
+  return
+    std::find_if(
+      annotations.begin(),
+      annotations.end(),
+      [&annotation_type_name](
+        const java_bytecode_parse_treet::annotationt &annotation)
+      {
+        if(annotation.type.id() != ID_pointer)
+          return false;
+        irept type = annotation.type.get_sub()[0];
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: `typet type = annotation.type.subtype()`?

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 12]
```diff
@@ -592,8 +711,41 @@ void java_bytecode_convert_classt::add_array_types(symbol_tablet &symbol_table)
...
+        if(type.id() == ID_symbol)
+        {
+          return
+            type.get_named_sub()[ID_identifier].id() == annotation_type_name;
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: `to_symbol_type(type).get_identifier()`

File: src/java_bytecode/java_bytecode_language.cpp [hunk 13]
```diff
@@ -44,7 +44,7 @@ Author: Daniel Kroening, kroening@kroening.com
 void java_bytecode_languaget::get_language_options(const cmdlinet &cmd)
 {
   assume_inputs_non_null=cmd.isset("java-assume-inputs-non-null");
-  java_class_loader.use_core_models=!cmd.isset("no-core-models");
+  java_class_loader.set_use_core_models(!cmd.isset("no-core-models"));
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_language.cpp:46]: :+1: but would have been better in a separate commit for future reference

**Review by thk123 — COMMENTED** [review]
Reviewed up to file: `src/java_bytecode/java_bytecode_language.cpp` then eyes started to glaze over

File: src/java_bytecode/java_bytecode_convert_class.cpp (hunk 1 above)
Comment (NathanJPhillips) [reply src/java_bytecode/java_bytecode_convert_class.cpp]: You should definitely put your overlay classes in a folder/jar that you add to the end of your classpath. We depend on the order of the classpath already as we currently only load the first instance seen.

File: src/java_bytecode/java_bytecode_convert_class.cpp (hunk 3 above)
Comment (NathanJPhillips) [reply src/java_bytecode/java_bytecode_convert_class.cpp]: In many languages I'd do this, but in C++ it doesn't really reduce the code and I'm not convinced it would do the `std::cref` correctly.

File: src/java_bytecode/java_bytecode_convert_class.cpp (hunk 10 above)
Comment (NathanJPhillips) [reply src/java_bytecode/java_bytecode_convert_class.cpp:540]: Not an invariant because it depends on user data rather than on CBMC programme logic.

File: src/java_bytecode/java_bytecode_language.cpp (hunk 13 above)
Comment (NathanJPhillips) [reply src/java_bytecode/java_bytecode_language.cpp:46]: Was only necessary because we need to clear a cache now when this is set so specific to this change.

**Comment by NathanJPhillips** [comment]
@thk123 - many comments added and your last two style suggestions applied.

File: src/java_bytecode/java_bytecode_language.cpp [hunk 14]
```diff
@@ -478,53 +478,53 @@ bool java_bytecode_languaget::typecheck(
...
+  for(const auto &c : java_class_loader.get_class_with_overlays_map())
   {
-    if(c.second.parsed_class.name.empty())
+    if(c.second.front().parsed_class.name.empty())
       continue;
     try
     {
       mark_java_implicitly_generic_class_type(
-        c.second.parsed_class.name, symbol_table);
+        c.second.front().parsed_class.name, symbol_table);
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_language.cpp:623]: Is there a reason we only care about the first parse tree in this case? Note an implicitly generic class means something like:

```java
class Outer<T> {
 class Inner { 
    T field;
  }
}
```

Inner isn't explicitly generic (has no generic parameters), but clearly is in some sense a generic type as one of its fields is of type `T`. I suspect you are better placed to know if it is important for any overlayed class parse trees to be correctly, but my feeling is it probably should be (the parse essentially means that the type of `field` is correctly updated to be `Outer::T`)

**Review by thk123 — COMMENTED** [review]
Reviewed the other half. Found it quite hard to review the changes to java_class_loaded, I think it would have helped to split some of into separate commits (e.g. swapping some of the iterator based for loops to range based for loops) 

I also note there are no tests - I really think a few simple uses cases of overlays would go along way to helping reviewers understand exactly what is going on (I'm sure you have your own tests in a separate repo but it'd be great if we can extract some kernel of them into this PR).

File: src/java_bytecode/java_bytecode_language.cpp (hunk 14 above)
Comment (NathanJPhillips) [reply src/java_bytecode/java_bytecode_language.cpp:623]: We care about them all, but they all have the same name. So `c.second.front().parsed_class.name` should be the same as `c.second[x].parsed_class.name`. At least that was my reasoning. I may be wrong.

Comment (thk123) [reply src/java_bytecode/java_bytecode_language.cpp:623]: Ah sorry obviously didn't read it carefully enough - you are of course correct here.

**Comment by NathanJPhillips** [comment]
@thk123 - added some tests, hope this helps.

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 15]
```diff
@@ -46,41 +46,105 @@ class java_bytecode_convert_classt:public messaget
...
     add_array_types(symbol_table);
 
-    bool loading_success=parse_tree.loading_successful;
-    if(loading_success)
-      convert(parse_tree.parsed_class);
+    // Ignore all parse trees that failed to load
+    std::list<std::reference_wrapper<const classt>>
+      loaded_parse_trees;
+    for(const java_bytecode_parse_treet &parse_tree : parse_trees)
+      if(parse_tree.loading_successful)
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: Since for loop body is over multiple lines, please wrap in braces

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 16]
```diff
@@ -46,41 +46,105 @@ class java_bytecode_convert_classt:public messaget
...
+        loaded_parse_trees.push_back(std::cref(parse_tree.parsed_class));
+    auto parse_tree_it = loaded_parse_trees.begin();
+    // Check that the first class implementation isn't an overlay
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]:
```
// Check that the first class implementation is an overlay
```
Seems more accurate, since it also checks the second isn't an overlay if the first is right? 

I'd be tempted to write this differently (e.g. `std::find` the first non-overlay class and then iterate up to it with the warnings) as this feels backwards to me (the expected condition is that the loop breaks on the first element). But I think that is just a preference so definitely ignore if you disagree.

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 17]
```diff
@@ -46,41 +46,105 @@ class java_bytecode_convert_classt:public messaget
...
+        << eom;
+      ++parse_tree_it;
+    }
+    bool loading_success = false;
+    if(parse_tree_it != loaded_parse_trees.end())
+    {
+      const classt &parsed_class = *parse_tree_it;
+      // Collect overlay classes
+      overlay_classest overlay_classes;
+      for(++parse_tree_it;
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: Personally would avoid variable reuse here and do something like `const auto overlay_class_it = std::next(parse_tree_it)` so you have a more "normal" looking for loop, but again definitely personal preference if you disagree.

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 18]
```diff
@@ -46,41 +46,105 @@ class java_bytecode_convert_classt:public messaget
...
+        parse_tree_it != loaded_parse_trees.end();
+        ++parse_tree_it)
+      {
+        const classt &overlay_class = *parse_tree_it;
+        // Ignore non-initial classes that aren't overlays
+        if(is_overlay_class(overlay_class))
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: I prefer making the expected route really clear at the expense of the error states being a bit more cumbersome (in this case, I'd use `std::filter` to make it clear you are filtering out the non-overlay classes, but again clearly a personal preference

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 19]
```diff
@@ -160,14 +224,11 @@ static optionalt<std::string> extract_generic_interface_reference(
   return {};
 }
 
-void java_bytecode_convert_classt::convert(const classt &c)
+void java_bytecode_convert_classt::convert(
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: Please update the (non-existant doxygen docs) to at least explain your parameter

File: src/java_bytecode/java_bytecode_convert_class.cpp (hunk 5 above)
Comment (thk123) [reply src/java_bytecode/java_bytecode_convert_class.cpp]: Not sure if you disagree or didn't see this, so just flagging it up.

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 20]
```diff
@@ -315,20 +376,111 @@ void java_bytecode_convert_classt::convert(const classt &c)
...
+  {
+    for(const methodt &method : overlay_class.get().methods)
+    {
+      const irep_idt method_identifier =
+        id2string(qualified_classname) + "." + id2string(method.name)
+          + ":" + method.descriptor;
+      if(method_bytecode.contains_method(method_identifier))
+      {
+        // This method has already been discovered while parsing an overlay
+        // class later in the classpath
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: Earlier rather than later surely?

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 21]
```diff
@@ -315,20 +376,111 @@ void java_bytecode_convert_classt::convert(const classt &c)
...
+        // If that definition is an overlay then we simply don't load this
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: I think it is confusing to have a comment that says if x directly above `if !x` which is what we have here right? I'd invert the comment:

```
// Warn the user if the existing definition is not an overlay
// otherwise we simply don't load this definition
```
Again, shouldn't this be `later definition` in your original comment?

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 22]
```diff
@@ -315,20 +376,111 @@ void java_bytecode_convert_classt::convert(const classt &c)
...
+      {
+        // This method was defined in a previous class definition without
+        // being marked as an overlay method
+        warning()
+          << "Method " << method_identifier
+          << " exists in an overlay class without being marked as an overlay "
+            "and also exists in the real class"
+          << eom;
+      }
+      continue;
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: So an method in an overlay class overwrites methods in the original class even if not marked as an overlay method (with only a warning) - I'd expect the behaviour in this case would be to not load the overlay method TBH.

File: src/java_bytecode/java_bytecode_language.cpp [hunk 23]
```diff
@@ -673,10 +671,11 @@ bool java_bytecode_languaget::typecheck(
   {
     journalling_symbol_tablet symbol_table_journal =
       journalling_symbol_tablet::wrap(symbol_table);
-    for(const auto &c : java_class_loader.class_map)
+    for(auto &class_to_trees : java_class_loader.get_class_with_overlays_map())
     {
-      create_stub_global_symbols(
-        c.second, symbol_table_journal, class_hierarchy, *this);
+      for(const java_bytecode_parse_treet &parse_tree : class_to_trees.second)
+        create_stub_global_symbols(
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_language.cpp]: For loop body now over multiple lines please wrap in braces

File: src/java_bytecode/java_class_loader.cpp [hunk 24]
```diff
@@ -57,214 +54,190 @@ java_bytecode_parse_treet &java_class_loadert::operator()(
...
-
-  // This does not read from the jar file but from the jar_filet object
-  // as we've just created it
-  read_jar_file(class_loader_limit, core_models);
-  return
-    get_class_file(class_loader_limit, class_name, core_models, parse_tree);
+  java_bytecode_parse_treet parse_tree;
+  std::istringstream istream(data);
+  if(java_bytecode_parse(istream, parse_tree, get_message_handler()))
+    return optionalt<java_bytecode_parse_treet>();
```
Comment (thk123) [inline src/java_bytecode/java_class_loader.cpp]: Prefer `return {}` for the optional without a value for consistency

File: regression/cbmc-java/overlay-class/annotations/com/diffblue/OverlayMethodImplementation.java [hunk 25]
```diff
@@ -0,0 +1,4 @@
+package com.diffblue;
+
+public @interface OverlayMethodImplementation {
```
Comment (thk123) [inline regression/cbmc-java/overlay-class/annotations/com/diffblue/OverlayMethodImplementation.java:3]: You might consider pulling this into a separate repo so you don't have to keep providing implementations of these OverlayMethod/Class attributes.

File: regression/cbmc-java/overlay-class/unmarked-overlay/Test.java [hunk 26]
```diff
@@ -0,0 +1,10 @@
+import com.diffblue.OverlayClassImplementation;
+import com.diffblue.OverlayMethodImplementation;
+
+public class Test
+{
+  @OverlayMethodImplementation
+  public static void main(String[] args)
+  {
```
Comment (thk123) [inline regression/cbmc-java/overlay-class/unmarked-overlay/Test.java:8]: To make this test more robust I'd probably put an `assert(true)` here an an `assert(false)` in the other original one.

**Review by thk123 — COMMENTED** [review]
Test could do with overlaid fields, some general feedback, don't think anything blocking (other than elaborating the tests)

File: src/java_bytecode/java_bytecode_convert_class.cpp (hunk 5 above)
Comment (NathanJPhillips) [reply src/java_bytecode/java_bytecode_convert_class.cpp]: Fields like `@class_identifier` will always be duplicated and can't be marked as overlays. I am happy that the final overlay gets to specify the final type/value of the fields, though we should probably warn if the types don't match.

**Comment by NathanJPhillips** [comment]
Thank you for the great review, @thk123 - as always you make my code better.

File: src/java_bytecode/java_bytecode_convert_class.cpp (hunk 19 above)
Comment (NathanJPhillips) [reply src/java_bytecode/java_bytecode_convert_class.cpp]: Done at the declaration.

**Comment by NathanJPhillips** [comment]
Had to add ability to `test.pl` to run different tests on different OS's to make my tests work - this probably needs a re-review.

**Comment by tautschnig** [comment]
Do you have to put the class files in subdirectories? It seems that's the only reason you need OS-specific test.desc files?!

File: src/java_bytecode/java_bytecode_convert_class.cpp (hunk 19 above)
Comment (thk123) [reply src/java_bytecode/java_bytecode_convert_class.cpp]: As much as it pains me as clearly the deceleration is the correct place - the coding guidelines say to do docs the implementation which is therefore where people look for this - could you move it?

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 27]
```diff
@@ -317,20 +393,114 @@ void java_bytecode_convert_classt::convert(const classt &c)
...
+  for(auto overlay_class : overlay_classes)
+  {
+    for(const methodt &method : overlay_class.get().methods)
+    {
+      const irep_idt method_identifier =
+        qualified_classname + "." + id2string(method.name)
+          + ":" + method.descriptor;
+      if(method_bytecode.contains_method(method_identifier))
+      {
+        // This method has already been discovered while parsing an overlay
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: Sorry I realise you've tried to explain this more thoroughly, but I still don't get why `overlay_methods.count(method_identifier) == 0` is equivalent to the overlay method existing later in the classpath? Surely the first method on the first overlay class we see, overlay_methods will be empty and therefore this will be true. Perhaps I don't know what `method_bytecode` is?

File: regression/cbmc-java/overlay-class/Test.java [hunk 28]
```diff
@@ -0,0 +1,13 @@
+public class Test
+{
+  public int x;
+
+  public static void main(String[] args)
+  {
+  }
```
Comment (thk123) [inline regression/cbmc-java/overlay-class/Test.java]: `assert(false)` here?

**Review by thk123 — APPROVED** [review]
Thanks for the adjustments, I'm still a little confused by that warning, perhaps we can talk it out IRL?

**Comment by NathanJPhillips** [comment]
@tautschnig 

> Do you have to put the class files in subdirectories? It seems that's the only reason you need OS-specific test.desc files?!

It's a good point, the reason I have sub-directories is because I need to put multiple folders on the class path. Now you ask the question the obvious way around this would have been to compile each one into a jar and run the test against those. It's not a bad idea to have a regression test that exercises the ability to use folders on the class path though.

**Comment by tautschnig** [comment]
> It's not a bad idea to have a regression test that exercises the ability to use folders on the class path though.

I agree, but I just wanted to understand whether this is essential here or not. We should seek to make tests as portable as possible. Arguably there also need to be tests that exercise the platform-specific bits, but ideally we'd keep them to a minimum. Making tests portable often is a bit of extra work, but I have always found work towards portability to be highly worthwhile.

File: src/java_bytecode/java_bytecode_convert_class.cpp (hunk 27 above)
Comment (NathanJPhillips) [reply src/java_bytecode/java_bytecode_convert_class.cpp]: Added more comment. If it's still not clear then come and ask me.

**Comment by NathanJPhillips** [comment]
@tautschnig With a bit more thought, even the idea of using jar files wouldn't make this portable as the class path delimiter is different on Windows too. I think we do need to be able to have different tests on different platforms to support testing anything to do with the class path. It's amazing that none of our tests have run into this before though.

**Comment by tautschnig** [comment]
The only place with limited flexibility is the command line (line 3 of *.desc). In all other places you'd just use regular expressions to cope with the different delimiters. I don't understand why Java doesn't permit multiple uses of `-cp` or `-classpath` and instead mandates that just one such is passed in a manner that is not compatible across platforms. 

That said: could you try whether the following would actually work:
```
--classpath '.;annotations;correct-overlay;:.:annotations:correct-overlay'
```
I believe Unix-type operating systems would discard the first one (the path just doesn't exist), and Windows would discard the second one. It seems to work in my experiments.

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 29]
```diff
@@ -317,20 +393,115 @@ void java_bytecode_convert_classt::convert(const classt &c)
...
+      const irep_idt method_identifier =
+        qualified_classname + "." + id2string(method.name)
+          + ":" + method.descriptor;
+      if(method_bytecode.contains_method(method_identifier))
+      {
+        // This method has already been discovered and added to method_bytecode
+        // while parsing an overlay class that appears later in the classpath
+        // (we're working backwards)
+        // Warn the user if the definition already found was not an overlay,
+        // otherwise simply don't load this earlier definition
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp:477]: Got it! :+1:

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 30]
```diff
@@ -317,20 +393,115 @@ void java_bytecode_convert_classt::convert(const classt &c)
...
+  // methods defined there
+  for(auto overlay_class : overlay_classes)
+  {
+    for(const auto &field : overlay_class.get().fields)
+    {
+      std::string field_id = qualified_classname + "." + id2string(field.name);
+      if(check_field_exists(field, field_id, fields))
+      {
+        // TODO: This could be a warning if the types match
+        error()
```
Comment (smowton) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: If this is an error it should throw (not an invariant, as this is bad user input). At present it uses the error() stream but behaves as a warning.

File: src/java_bytecode/java_bytecode_convert_class.cpp [hunk 31]
```diff
@@ -317,20 +393,115 @@ void java_bytecode_convert_classt::convert(const classt &c)
...
+      // If that definition is an overlay then we simply don't load this
+      // original definition and we remove it from the list of overlays
+      if(overlay_methods.erase(method_identifier) == 0)
+      {
+        // This method was defined in a previous class definition without
+        // being marked as an overlay method
+        warning()
+          << "Method " << method_identifier
+          << " exists in an overlay class without being marked as an overlay "
+            "and also exists in the real class"
```
Comment (smowton) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: Suggest "underlying class"

File: src/util/read_only_map_wrapper.h [hunk 32]
```diff
@@ -0,0 +1,35 @@
...
+
+/// \file
+/// A wrapper for maps that gives read-write access to elements but without
+/// allowing addition or removal of elements
+
+#ifndef CPROVER_UTIL_READ_ONLY_MAP_WRAPPER_H
+#define CPROVER_UTIL_READ_ONLY_MAP_WRAPPER_H
+
+template <typename mapt>
+class read_only_map_wrappert
```
Comment (smowton) [inline src/util/read_only_map_wrapper.h]: Rename `map_fixed_keys_wrappert` or something

File: src/util/read_only_map_wrapper.h [hunk 33]
```diff
@@ -0,0 +1,35 @@
...
+{
+private:
+  mapt &map;
+
+public:
+  explicit read_only_map_wrappert(mapt &map) : map(map)
+  {
+  }
+
+  typename mapt::iterator begin()
```
Comment (smowton) [inline src/util/read_only_map_wrapper.h]: Add .at, operator[] (but the latter throws if not found)

**Review by smowton — APPROVED** [review]
Some nitpicks, otherwise lgtm.

**Comment by tautschnig** [comment]
@NathanJPhillips Would you mind giving
```
--classpath '.;annotations;correct-overlay;:.:annotations:correct-overlay'
```
a spin? Not adding more features to `test.pl` would seem worthwhile (and cross-platform code is even more so).

**Comment by smowton** [comment]
@tautschnig if we didn't already have a clean solution that'd be an ingenious hack to dodge the work, but given we've got one (and platform-specific tests are surely needed for some other things, e.g. compiler-specific options to pass to goto-cc without producing an error on the other platform?), I suggest we go with the solution proposed.

**Comment by tautschnig** [comment]
Would it be possible to factor the `test.pl` changes out in a separate PR to discuss properly?

**Comment by NathanJPhillips** [comment]
https://github.com/diffblue/cbmc/pull/1929

**Comment by smowton** [comment]
@tautschnig @NathanJPhillips here's an example of the shell-script solution: https://github.com/smowton/cbmc/commit/bf075a3d6fc6e695f7b188fb4821e0915e93ec99

Shall we just go with that for now, since it's simple, non-invasive and gets this PR out of its current quagmire?

**Comment by tautschnig** [comment]
It's almost scary that you can actually use backticks in the command line arguments, but indeed a nice hack. I'm not going to argue against it.

**Review by smowton — APPROVED** [review]
